### PR TITLE
Change behaviour of Garbage Collector 

### PR DESCRIPTION
**Pull request by piotrnosek**
Only remove AggregateCollectionStates which don't have an existing corresponding controller (e.g. Deployment).

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/model/cluster.go [hunk 1]
```diff
@@ -348,10 +349,15 @@ func (cluster *ClusterState) findOrCreateAggregateContainerState(containerID Con
 	return aggregateContainerState
 }
 
-func (cluster *ClusterState) garbageCollectAggregateCollectionStates(now time.Time) {
+// GarbageCollectAggregateCollectionStates removes obsolete AggregateCollectionStates from the ClusterState.
+// AggregateCollectionState is obsolete in following situations:
+// 1) It has no samples and there are no more active pods that can contribute,
+// 2) The last sample is too old to give meaningful recommendation (>8 days),
+// 3) There are no samples and the aggregate state was created >8 days ago.
+func (cluster *ClusterState) garbageCollectAggregateCollectionStates(now time.Time, controllerFetcher controllerfetcher.ControllerFetcher) {
```
[the remark this answers is not in the document]
Comment (piotrnosek) [reply vertical-pod-autoscaler/pkg/recommender/model/cluster.go:357]: Actually all points are still true, just the definition of active pod changed. Before an inactive pod would be a pod that is in a terminal state (succeeded/failed). Right now, an inactive pod is a pod which is both in a terminal state and doesn't have an existing controller. I've added a comment to reflect that.

This doesn't change the logic for old samples (>8 days old) and old aggregates.

[the remark this answers is not in the document]
Comment (jbartosik) [reply vertical-pod-autoscaler/pkg/recommender/model/cluster.go:357]: Saying pod is active when its phase is not one of {PodSucceeded, PodFailed} makes sense to me.

Saying pod is active when its phase is not one of {PodSucceeded, PodFailed} or there is a controller for it looks unintuitive to me.

Please:
- update this change to keep previous definition of active and add having a controller as a separate condition, or
- pick a new word for the concept "has a controller or isn't in a terminal phase".

**Review by jbartosik — CHANGES_REQUESTED** [review]

**Comment by jbartosik** [comment]
@piotrnosek please rebase this PR on top of current master, it looks like it has a lot of changes that shouldn't be here.

**Comment by piotrnosek** [comment]
Closing this PR due to running into rebase hell with git, desired changes are on a separate PR: https://github.com/kubernetes/autoscaler/pull/4488.

**Pull request closed without merging**